nixos: loader: added generic config.boot.loader.timeout option

bobvanderlinden opened the pull request:
This adds the option `config.boot.loader.timeout` that can be used to set the timeout of grub as well as gummiboot. You can still override this setting for each specific loader by setting `config.boot.loader.grub.timeout` or `config.boot.loader.gummiboot.timeout`, but they will all use `config.boot.loader.timeout` by default.

Do note that some conversions are happening on Grub and Gummiboots side, since the value of `config.boot.loader.timeout` behaves differently from `config.boot.loader.grub.timeout` or `config.boot.loader.gummiboot.timeout`... though these two bootloaders timeout behave differently from eachother as well.

`null` is no timeout. `0` is instant booting. Any positive integer is the timeout before automatically booting in seconds. Negative integers are undefined behavior.

Lastly, I haven't touched references to `config.boot.loader.grub.timeout`, but I'd imagine most of these will need to be changed to avoid confusing behavior.

I have little experience with the Nix language, so any suggestions improving on the (nested) `if ... then ... else ...` would be appreciated.

bobvanderlinden committed:
nixos: loader: added generic config.boot.loader.timeout option

timeout options of grub and gummiboot will inherit the value of this
option by default.

lethalman commented:
Looks ok to me. Unfortunately we cannot do deprecation using rename.nix because it doesn't support one option deprecating two. cc @edolstra

bobvanderlinden commented:
The specific options aren't really deprecated though. It would often be better to use the new generic option, but there probably are use-cases where using the option specific to the bootloader is good as well (I think).

7c6f434c commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -196,7 +196,7 @@ in
 
       timeout = mkOption {
-        default = 5;
+        default = if (config.boot.loader.timeout != null) then config.boot.loader.timeout else -1;
```
So you change the default from the previous one to some unified default?

lethalman replied:
Yes.

bobvanderlinden replied:
This is also why I thought deprecating isn't a good idea. It's all still valid to use, but it might have non-obvious consequences. Is it better to just deprecate `config.boot.loader.grub.timeout`?

lethalman replied:
As said we can't deprecate all of them because it's not supported in
rename.nix. I'd say it's ok as is.

On Tue, Jan 20, 2015 at 12:10 AM, bobvanderlinden <notifications@github.com>
wrote:

> In nixos/modules/system/boot/loader/grub/grub.nix
> <https://github.com/NixOS/nixpkgs/pull/5772#discussion-diff-23193233>:
>
> > @@ -196,7 +196,7 @@ in
> >        };
> >
> >        timeout = mkOption {
> > -        default = 5;
> > +        default = if (config.boot.loader.timeout != null) then config.boot.loader.timeout else -1;
>
> This is also why I thought deprecating isn't a good idea. It's all still
> valid to use, but it might have non-obvious consequences. Is it better to
> just deprecate config.boot.loader.grub.timeout?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/5772/files#r23193233>.
>



-- 
NixOS Linux <http://nixos.org>

lethalman commented on nixos/modules/system/boot/loader/gummiboot/gummiboot.nix:
```diff
@@ -31,7 +31,9 @@ in {
 
     timeout = mkOption {
-      default = null;
+      default = if (config.boot.loader.timeout != null) then
```
Was going to test this PR, however here does it mean a timeout of either null or 0 means no entry in gummiboot? Note that unsetting the timeout in gummiboot means the entry is executed immediately, afaik, which is different than the description of `boot.loader.timeout`.

bobvanderlinden committed:
nixos: gummiboot: change default to 1000 if boot.loader.timeout is null

When gummiboot.timeout == null, the menu will still be skipped.
When gummiboot.timeout == 0, the menu will also be skipped.
The only way to show the menu 'indefinitely' is to show it a long time.

bobvanderlinden commented:
`gummiboot.timeout` seems to behave differently than described. I tried to set it to `null`, `-1` and `0`, but all of these resulted in the menu being skipped instead of shown indefinitely. I'm not entirely sure why, because the code http://cgit.freedesktop.org/gummiboot/tree/src/efi/gummiboot.c#n583 seems to suggest otherwise. Anyway, Lethalman suggested this was another issue entirely, so will not be fixed here.

lethalman commented:
Merged manually in a239775 . I changed the gummiboot timeout from 1000 to 10000. Thanks.

bobvanderlinden commented:
Thanks! I'll integrate this into the syslinux PR.

On Fri, Jan 30, 2015 at 10:41 AM, lethalman <notifications@github.com>
wrote:

> Merged manually in a239775
> <https://github.com/NixOS/nixpkgs/commit/a239775488722bc8325e9d5bdfefd20bfab38164>
> . I changed the gummiboot timeout from 1000 to 10000. Thanks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/5772#issuecomment-72176450>.
>
